go/vt/vttablet/tabletserver/tabletserver.go

```diff
@@ -1331,38 +1331,61 @@ func (tsv *TabletServer) convertError(ctx context.Context, sql string, bindVaria
 	errstr := err.Error()
 	errnum := sqlErr.Number()
 	sqlState := sqlErr.SQLState()
+
 	switch errnum {
-	case mysql.EROptionPreventsStatement:
-		// Special-case this error code. It's probably because
-		// there was a failover and there are old clients still connected.
-		if strings.Contains(errstr, "read-only") {
```

Comment:
Which error did we return in the implicit "else" branch before your change?

Do I see it correctly, that you're always returning FAILED_PRECONDITION now if the MySQL error code is EROptionPreventsStatement?

I think this may be a regression.

It looks like 1290 is another MySQL code which was repurposed by the internal Google MySQL patches? When I google for "mysql error 1290", results always talk about "secure-file-priv" and never about running in read-only mode. Pinging @pivanof as well here.

If that's the case, I think we need to keep this if and add an else here like you did for access denied below (another MySQL error repurposed by Google patches).

You may wonder why this is relevant: The vtgate buffer gets triggered based on the Vitess error code (FAILED_PRECONDITION) and the MySQL error number. See: https://github.com/youtube/vitess/blob/master/go/vt/vtgate/buffer/buffer.go#L255 Since we're stripping the MySQL error message to protect PII, we can no longer check for a "read-only" string in the error message in vtgate. As a consequence, we may detect these "secure-file-priv" problems, the original purpose of the MySQL error, as failover and start buffering.

Reply:
If you search for "mysql read_only error", the first link to StackOverflow quotes: "General error: 1290 The MySQL server is running with the --read-only option so it cannot execute this statement". We didn't repurpose 1290 for read_only -- it's the native error code.

Reply:
@michael-berlin you can see the original code here: https://github.com/youtube/vitess/blob/master/go/vt/vttablet/tabletserver/tabletserver.go#L1338

The default value for errCode is `UNKNOWN`. We returned `FAILED_PRECONDITION` only for the read-only case.  My change makes us return `FAILED_PRECONDITION` in all cases. I think this makes sense.

`ERSpecifiedAccessDenied` (1227) was repurposed for google, and I kept that logic in tact.

Reply:
Btw, the diff lined things up weirdly, because I also reordered the blocks.  See that `EROptionPreventsStatement` still returns `RESOURCE_EXHAUSTED` by searching for `EROptionPreventsStatement` in the diff and see which switch case it matches against.

Reply:
The code you linked to in buffer.go also expected `FAILED_PRECONDITION`

Reply:
Sorry, I wasn't clear enough here.

Please readd the `if` here i.e. if read-only, then `FAILED_PRECONDITION`, otherwise any other error code (which one makes sense?).

If we don't do that, then non read-only errors can trigger the vtgate buffer. This is bad because the purpose of the vtgate buffer is to let MASTER requests survive master failovers by detecting them, stalling pending **and** future requests and retrying them when the failover is done. In case of a non read-only error, we will not observe the failover end and instead buffering will stop automatically after the [timeout (20 seconds by default)](https://github.com/youtube/vitess/blob/master/go/vt/vtgate/buffer/flags.go#L35).

When do non read-only 1290 errors occur? Apparently when you send LOAD queries and try to access a file outside the path specified by MySQL's `--secure_file_priv`. That means the current version would allow people to provoke buffering and stalling requests although no failover is in progress. (Note that buffering is turned off by default and must be explicitly enabled.)

If you're saying, we should fix the buffering trigger instead, I agree with you. The problem is that the MySQL error message itself is stripped at that point i.e. vtgate cannot find out if it's a read-only error or not. Instead, we only see 1290 and `FAILED_PRECONDITION`. Therefore, we must not see `FAILED_PRECONDITION` if it wasn't a read-only error.

In general, we're confined by the Vitess error space here. We do this on purpose to be consistent with the Google canonical error codes, i.e. we do not want to add a new error code which says "this error should trigger buffering". As an alternative, I think we can add code in vttablet which would determine an error as a buffer trigger and add a special string in the error message as marker after the MySQL error message was possibly stripped. But that's a different pull request for another day ;-)

Note that even if we had not this weird dependency on the buffering trigger code, I think `FAILED_PRECONDITION` for all 1290 errors would not be correct. `FAILED_PRECONDITION` allows to restart things at a higher level e.g. a transaction from the beginning again. We mostly use it to signal vtgate that a REPLICA vttablet shut down and vtgate should try a different REPLICA instead. This is not applicable for the `secure_file_priv` error: The user tries to access data which it isn't allowed to and this shouldn't change when they use a different task.

Does all this make more sense now?

If so, can you recommend an error code for the non read-only errors?

Reply:
Thanks, that makes sense. I'll limit to just the read-only errors, and keep the rest as `UNKNOWN` for now